site_tools/image_deployer.py

```diff
@@ -0,0 +1,297 @@
+        logging.debug(f"Image {image_disk_name} current value: {image_current_value}")
+        if image_current_value is not None and image_current_value == current:
+            logging.info(f"Image {image_disk_name} is already current.")
+            return False
+    return True
+
+
+def download_object_to_file(storage_url, path, file_name):
+    url = f"{storage_url}/{path}/{file_name}"
+    response = requests.get(url, stream=True)
+
+    if response.status_code != 200:
+        raise Exception(f"Error downloading object {file_name}: {response.content}")
+
+    with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+        for chunk in response.iter_content(chunk_size=65536):
+            temp_file.write(chunk)
+            temp_file.flush()
+    logging.debug(f"Downloaded object to {temp_file.name}.")
+    return temp_file
+
+
+def upload_image_to_glance(image_connection,
+                           image_prefix,
+                           image_disk_name,
+                           image_file_name,
+                           disk_format,
+                           manifest_data):
+    image_prefix_name = image_prefix + image_disk_name
+
+    # TODO: need admin creds to make visibility public instead of private
+    # right now I am using app creds that just have a member role
+    logging.info(f"Uploading image {image_disk_name} to Glance.")
+    with open(image_file_name, "rb") as image_data:
+        new_image = image_connection.create_image(name=image_prefix_name,
+                                                  disk_format=disk_format,
+                                                  container_format="bare",
+                                                  visibility="private",
+                                                  data=image_data,
+                                                  **manifest_data)
+    logging.info(f"Uploaded image {new_image.name}.")
+    return new_image
+
+
+def archive_image(image_connection, image_disk_name, new_image):
+    # TODO: if things go sideways in here we may be in a bad state, add more
+    # error handling.
+    existing_images = list(image_connection.image.images(name=image_disk_name))
+    # TODO: rename the old image: this should only ever be 1. assert that instead?
+    if len(existing_images) > 0:
+        logging.info(f"Renaming existing image {image_disk_name}.")
+        archive_date = datetime.datetime.now().strftime("%Y%m%d_%H%M")
+        for existing_image in existing_images:
+            image_connection.image.update_image(
+                existing_image.id,
+                name=f"{image_disk_name}_{archive_date}"
+            )
+            logging.info(f"Renamed image {existing_image.name} " + \
+                         f"to {existing_image.name}_{archive_date}.")
+    image_connection.image.update_image(new_image.id, name=image_disk_name)
+    logging.info(f"Renamed image {new_image.id} to {image_disk_name}.")
```

Comment:
I think the logic here is something like: "ensure exactly 1 image is both public and has name image_disk_name"

something like:
0. upload the new image with a temp name, and private visibility
1. list images having visibility=public and name=image_disk_name
2a. if count=0, just promote our new image to public + this name
2b. if count=1, archive the existing image first, then promote the new one
2c. if count>1, we've somehow gotten into a bad state before running.

because most of the logic isn't actually about archiving, it might make more sense to have a separate method for "promote image", and it can just call archive image when needed

Reply:
updated to follow this logic